pkg/apis/rbac/types.go

```diff
@@ -0,0 +1,252 @@
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+// Authorization is calculated against
+// 1. all deny RoleBinding PolicyRules in the master namespace - short circuit on match
+// 2. all allow RoleBinding PolicyRules in the master namespace - short circuit on match
+// 3. all deny RoleBinding PolicyRules in the namespace - short circuit on match
+// 4. all allow RoleBinding PolicyRules in the namespace - short circuit on match
+// 5. deny by default
+
+const (
+	// PolicyName is the name of Policy
+	PolicyName     = "default"
+	APIGroupAll    = "*"
+	ResourceAll    = "*"
+	VerbAll        = "*"
+	NonResourceAll = "*"
+)
+
+// PolicyRule holds information that describes a policy rule, but does not contain information
+// about who the rule applies to or which namespace the rule applies to.
+type PolicyRule struct {
+	// Verbs is a list of Verbs that apply to ALL the ResourceKinds and AttributeRestrictions contained in this rule.  VerbAll represents all kinds.
+	Verbs sets.String
+	// AttributeRestrictions will vary depending on what the Authorizer/AuthorizationAttributeBuilder pair supports.
+	// If the Authorizer does not recognize how to handle the AttributeRestrictions, the Authorizer should report an error.
+	AttributeRestrictions runtime.Object
+	// APIGroups is the name of the APIGroup that contains the resources.  If this field is empty, then the core API Group is assumed.
+	// If multiple API groups are specified, any action requested against one of the enumerated resources in any API group will be allowed.
+	APIGroups []string
+	// Resources is a list of resources this rule applies to.  ResourceAll represents all resources.
+	Resources sets.String
+	// ResourceNames is an optional white list of names that the rule applies to.  An empty set means that everything is allowed.
+	ResourceNames sets.String
+	// NonResourceURLs is a set of partial urls that a user should have access to.  *s are allowed, but only as the full, final step in the path
+	// If an action is not a resource API request, then the URL is split on '/' and is checked against the NonResourceURLs to look for a match.
+	NonResourceURLs sets.String
+}
+
+// Role is a logical grouping of PolicyRules that can be referenced as a unit by RoleBindings.
+type Role struct {
+	unversioned.TypeMeta
+	// Standard object's metadata.
+	api.ObjectMeta
+
+	// Rules holds all the PolicyRules for this Role
+	Rules []PolicyRule
+}
+
+// RoleBinding references a Role, but not contain it.  It can reference any Role in the same namespace or in the global namespace.
+// It adds who information via Users and Groups and namespace information by which namespace it exists in.  RoleBindings in a given
+// namespace only have effect in that namespace (excepting the master namespace which has power in all namespaces).
+type RoleBinding struct {
+	unversioned.TypeMeta
+	api.ObjectMeta
+
+	// Subjects hold object references of to authorize with this rule
+	Subjects []api.ObjectReference
```

Comment:
> Yeah, I guess we'd have to add additional user and group objects to reference, though an Authorization package seems like an odd place to put that type.

We don't *have* to create API objects. If it would make us feel better about it, instead of an `ObjectReference` we could have a `Subject` type with identically named fields (Kind, Name, Namespace), that only accepted three kinds (User, Group, ServiceAccount), and only allowed namespace to be set for ServiceAccounts. 

> Perhaps it'd be best to stick with the username/groupname split, then change it if reference-able types are added later?

Having followed that path in our impl, it resulted in a slightly annoying API... the old fields still have to be included for compatibility, and the structured field has to be ignored if the old fields are sent in a request. For service accounts in particular, we also found that specifying the information in a structured way (`{"kind":"ServiceAccount","name":"foo","namespace":"bar"}`) was way nicer than making users incant the `system:serviceaccount:bar:foo` format